src/devices/bus/hp9825_optroms/hp9825_optrom.cpp

```diff
@@ -0,0 +1,184 @@
+// license:BSD-3-Clause
+// copyright-holders: F. Ulivi
+/*********************************************************************
+
+    hp9825_optrom.cpp
+
+    Optional ROMs for HP9825 systems
+
+*********************************************************************/
+
+#include "emu.h"
+#include "hp9825_optrom.h"
+#include "softlist.h"
+
+// Debugging
+//#define VERBOSE 1
+#include "logmacro.h"
+
+DEFINE_DEVICE_TYPE(HP9825_OPTROM_CART, hp9825_optrom_cart_device, "hp9825_optrom_cart", "HP9825 optional ROM cartridge")
+DEFINE_DEVICE_TYPE(HP9825_OPTROM_SLOT, hp9825_optrom_slot_device, "hp9825_optrom_slot", "HP9825 optional ROM slot")
+
+struct optrom_region_t {
+	offs_t m_start;
+	offs_t m_size;
+	const char *m_tag;
+};
+
+constexpr std::array<struct optrom_region_t , 8> region_tab =
+	{
+	 0x3000 , 0x400 , "rom3000" ,
+	 0x3400 , 0x400 , "rom3400" ,
+	 0x3800 , 0x400 , "rom3800" ,
+	 0x3c00 , 0x400 , "rom3c00" ,
+	 0x4000 , 0x400 , "rom4000" ,
+	 0x4400 , 0x800 , "rom4400" ,
+	 0x4c00 , 0x400 , "rom4c00" ,
+	 0x5c00 ,0x2000 , "rom5c00"
+	};
+
+// +-------------------------+
+// |hp9825_optrom_cart_device|
+// +-------------------------+
+hp9825_optrom_cart_device::hp9825_optrom_cart_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: hp9825_optrom_cart_device(mconfig , HP9825_OPTROM_CART , tag , owner , clock)
+{
+}
+
+hp9825_optrom_cart_device::hp9825_optrom_cart_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, type, tag, owner, clock)
+	, device_slot_card_interface(mconfig, *this)
+{
+}
+
+// +-------------------------+
+// |hp9825_optrom_slot_device|
+// +-------------------------+
+hp9825_optrom_slot_device::hp9825_optrom_slot_device(machine_config const &mconfig, char const *tag, device_t *owner)
+	: hp9825_optrom_slot_device(mconfig, tag, owner, (uint32_t)0)
+{
+	option_reset();
+	option_add_internal("rom", HP9825_OPTROM_CART);
+	set_default_option(nullptr);
+	set_fixed(false);
+}
+
+hp9825_optrom_slot_device::hp9825_optrom_slot_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, HP9825_OPTROM_SLOT, tag, owner, clock)
+	, device_image_interface(mconfig, *this)
+	, device_slot_interface(mconfig, *this)
+	, m_cart(nullptr)
+	, m_rom_limit(0xffffU)
+	, m_loaded_regions(0)
+	, m_space_r(nullptr)
+	, m_bank(*this , "rombank")
+{
+}
+
+hp9825_optrom_slot_device::~hp9825_optrom_slot_device()
+{
+}
+
+void hp9825_optrom_slot_device::set_rom_limit(offs_t rom_limit)
+{
+	m_rom_limit = rom_limit;
+}
```

Comment:
Is this supposed to be used at configuration time?  If it is, you should just inline it in the header.  It makes the compiled code more compact if it can be flattened and no unwind table entries need to be generated for things like this.